File: C/src/bfvm.c
/**
 * @file   bfvm.c
 * @brief  A virtual machine for the Brainfuck programming language
 * @author Ruan C. Keet
 * @date   25-10-2023
 */

#include "bfc.h"
#include "error.h"

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/* --- constants ------------------------------------------------------------*/

#define DATA_SIZE 30000

/* --- global variables -----------------------------------------------------*/

static u8 data[DATA_SIZE]; /* data array */
static u16 dp;             /* data ptr */
static size_t ip;          /* instruction ptr */

/* --- execution routines ---------------------------------------------------*/

static void addb(u8 val);
static void subb(u8 val);
static void addp(u16 val);
static void subp(u16 val);
static void write(void);
static void read(void);
static void jz(size_t line);
static void jmp(size_t line);

/* --- main routine ---------------------------------------------------------*/

/**
 * The entry point to the program.
 * 
 * @param[in] argc
 *      The amount of arguments passed via the CLI.
 * @param[in] argv
 *      The arguments passed via the CLI.
 * 
 * @return
 *      The exit status of the program.
 */
int main(int argc, char *argv[])
{
    clock_t start, end;
    f32 delta;
    ByteCode *code;

    code = compile(argc < 2 ? NULL : argv[1]);
    ip = 0;

    start = clock();
    while (code[ip].op != BFVM_END) {
        switch (code[ip].op) {
            case BFVM_ADDB: addb(code[ip].bval); break;
            case BFVM_SUBB: subb(code[ip].bval); break;
            case BFVM_ADDP: addp(code[ip].dval); break;
            case BFVM_SUBP: subp(code[ip].dval); break;
            case BFVM_WRITE: write(); break;
            case BFVM_READ: read(); break;
            case BFVM_JZ: jz(code[ip].line); break;
            case BFVM_JMP: jmp(code[ip].line); break;
            default: log_err("unknown opcode %d\n", code[ip].op);
        }
    }
    end = clock();

    delta = (f32)(end - start) / CLOCKS_PER_SEC;
    log_norm("time elapsed: %.5fms", delta * 1000.0f);

    free(code);
    return EXIT_SUCCESS;
}

/* --- execution routines ---------------------------------------------------*/

inline static void addb(u8 val)
{
    data[dp] += val;
    ip++;
}

inline static void subb(u8 val)
{
    data[dp] -= val;
    ip++;
}

inline static void addp(u16 val)
{
    dp += val;
    if (dp >= DATA_SIZE) {
        log_err("data pointer out of range");
    }
    
    ip++;
}

inline static void subp(u16 val)
{
    dp -= val;
    if (dp >= DATA_SIZE) {
        log_err("data pointer out of range");
    }

    ip++;
}

inline static void write(void)
{
    if (putchar(data[dp]) == EOF) {
        log_err("failed to output byte");
    }

    ip++;
}

inline static void read(void)
{
    i32 ch;
    if ((ch = fgetc(stdin)) == EOF) {
        log_err("failed to read byte");
    }

    data[dp] = (u8)ch;
    ip++;
}

inline static void jz(size_t line)
{
    ip = (data[dp] != 0x00) ? ip + 1 : line;
}

inline static void jmp(size_t line)
{
    ip = line;
}

File: Rust/src/bfvm.rs

use crate::bfc::{ByteCode, Compiler, OpCode};

/// Representation of the virtual machine data and object.
pub struct VirtualMachine {
    data: [u8; 30000],
    data_ptr: usize,
    byte_code: Vec<ByteCode>,
    instruction_ptr: usize
}

impl VirtualMachine {
    /// Creates a new instance of the `VirtualMachine`.
    /// 
    /// # Arguments
    /// 
    /// `filepath` - The path to a Brainf*ck source file.
    /// 
    /// # Returns
    /// 
    /// A new instance of the virtual machine.
    pub fn new(filepath: &str) -> Self {
        let mut compiler: Compiler = Compiler::new(filepath);

        VirtualMachine {
            data: [0; 30000],
            data_ptr: 0,
            byte_code: compiler.compile(),
            instruction_ptr: 0
        }
    }

    /// Runs the Brainf*ck program via the virtual machine.
    pub fn run(&mut self) {
        while self.byte_code[self.instruction_ptr].code != OpCode::End {
            let bc: ByteCode = self.byte_code[self.instruction_ptr];
            match bc.code {
                OpCode::AddB => self.add_byte(bc.operand as u8),
                OpCode::SubB => self.sub_byte(bc.operand as u8),
                OpCode::AddP => self.add_ptr(bc.operand),
                OpCode::SubP => self.sub_ptr(bc.operand),
                OpCode::Write => self.write(),
                OpCode::Read => self.read(),
                OpCode::Jz => self.jump_zero(bc.operand),
                OpCode::Jmp => self.jump(bc.operand),
                _ => debug_assert!(false)
            }
        }
    }

    /// Adds `val` to the byte at the current `data_ptr`.
    /// 
    /// # Arguments
    /// 
    /// `val` - The value to increment the byte with.
    #[inline(always)]
    fn add_byte(&mut self, val: u8) {
        self.data[self.data_ptr] = self.data[self.data_ptr].wrapping_add(val);
        self.instruction_ptr += 1;
    }

    /// Subtracts `val` from the byte at the current `data_ptr`.
    /// 
    /// # Arguments
    /// 
    /// `val` - The value to decrement the byte with.
    #[inline(always)]
    fn sub_byte(&mut self, val: u8) {
        self.data[self.data_ptr] = self.data[self.data_ptr].wrapping_sub(val);
        self.instruction_ptr += 1;
    }

    /// Adds `val` to the `data_ptr`.
    /// 
    /// # Arguments
    /// 
    /// `val` - The amount to shift the `data_ptr` up by.
    #[inline(always)]
    fn add_ptr(&mut self, val: usize) {
        self.data_ptr = self.data_ptr.wrapping_add(val);

        if self.data_ptr >= 30000 {
            eprintln!("bfvm: error: data pointer out of range");
            std::process::exit(1);
        }
        self.instruction_ptr += 1;
    }

    /// Subtracts `val` from the `data_ptr`.
    /// 
    /// # Arguments
    /// 
    /// `val` - The amount to shift the `data_ptr` down by.
    #[inline(always)]
    fn sub_ptr(&mut self, val: usize) {
        self.data_ptr = self.data_ptr.wrapping_sub(val);

        if self.data_ptr >= 30000 {
            eprintln!("bfvm: error: data pointer out of range");
            std::process::exit(1);
        }
        self.instruction_ptr += 1;
    }

    /// Writes the byte at `data_ptr` to the `stdout`.
    #[inline(always)]
    fn write(&mut self) {
        print!("{}", self.data[self.data_ptr] as char);
        self.instruction_ptr += 1;
    }

    /// Reads in a byte from the user via `stdin`.
    #[inline(always)]
    fn read(&mut self) {
        // TODO: Read a single byte in from stdin
        self.instruction_ptr += 1;
    }

    /// Jumps to the instruction at `line` when the byte at `data_ptr` is zero.
    /// 
    /// # Arguments
    /// 
    /// `line` - The line in the assembly instructions to jump to.
    #[inline(always)]
    fn jump_zero(&mut self, line: usize) {
        if self.data[self.data_ptr] != 0 {
            self.instruction_ptr += 1;
        } else {
            self.instruction_ptr = line;
        }
    }

    /// Jumps to the instruction at `line`.
    /// 
    /// # Arguments
    /// 
    /// `line` - The line to jump to.
    #[inline(always)]
    fn jump(&mut self, line: usize) {
        self.instruction_ptr = line;
    }
}